### PR TITLE
Libressl Adaptations

### PR DESCRIPTION
**Pull request by VlkrS**
* Define HAVE_CHACHA20_POLY1305 and HAVE_CHACHA20 for LibreSSL >= 3.7.0
* Remove a limitation that is not needed since LibreSSL >= 3.5.0 anymore (#7043)

### CODE REVIEW

**Comment by sverker** [comment]
Added a fixup commit to actually test for LibreSSL < 3.5.0, if that is the version where the EVP_MD_meth_* function were introduced, which it seems to be.

File: lib/crypto/c_src/otp_test_engine.c [hunk 1]
```diff
@@ -37,7 +37,7 @@
     PACKED_OPENSSL_VERSION(MAJ,MIN,FIX,('a'-1))
 
 #if OPENSSL_VERSION_NUMBER < PACKED_OPENSSL_VERSION_PLAIN(1,1,0) \
-    || LIBRESSL_VERSION_NUMBER < 0x3050000fL
```
Comment (sverker) [inline lib/crypto/c_src/otp_test_engine.c]: That's actually not needed due to a quirk/feature in the preprocessor causing undefined macros to be replaced with 0 in #if-conditions.

I thought of maybe (ab)using that to define all-in-one macros like this

```
#ifdef LIBRESSL_VERSION_NUMBER
#  define HAS_LIBRESSL LIBRESSL_VERSION_NUMBER
#else  /* assume OpenSSL */
#  define HAS_OPENSSL OPENSSL_VERSION_NUMBER
#endif
```
in order to write simpler conditions like
```
#if HAS_OPENSSL > PACKED_VERSION_NUMBER(1,1,0) || HAS_LIBRESSL > 0x3050000f
#  define HAS_SOME_FEATURE
#endif
```

File: lib/crypto/c_src/otp_test_engine.c [hunk 2]
```diff
@@ -37,7 +37,7 @@
     PACKED_OPENSSL_VERSION(MAJ,MIN,FIX,('a'-1))
 
 #if OPENSSL_VERSION_NUMBER < PACKED_OPENSSL_VERSION_PLAIN(1,1,0) \
-    || defined(LIBRESSL_VERSION_NUMBER)
+    || (defined(LIBRESSL_VERSION_NUMBER) && LIBRESSL_VERSION_NUMBER < 0x3050000fL)
```
Comment (sverker) [inline lib/crypto/c_src/otp_test_engine.c:40]: That's actually not needed due to a quirk/feature in the preprocessor causing undefined macros to be replaced with 0 in #if-conditions.

I thought of maybe (ab)using that to define all-in-one macros like this

```
#ifdef LIBRESSL_VERSION_NUMBER
#  define HAS_LIBRESSL LIBRESSL_VERSION_NUMBER
#else  /* assume OpenSSL */
#  define HAS_OPENSSL OPENSSL_VERSION_NUMBER
#endif
```
in order to write simpler conditions like
```
#if HAS_OPENSSL > PACKED_VERSION_NUMBER(1,1,0) || HAS_LIBRESSL > 0x3050000f
#  define HAS_SOME_FEATURE
#endif
```

Comment (VlkrS) [reply lib/crypto/c_src/otp_test_engine.c:40]: I did see a CI build failure due to LIBRESSL_VERSION_NUMBER being undefined, which prompted me to propose that fix

Comment (sverker) [reply lib/crypto/c_src/otp_test_engine.c:40]: I take that back about abusing undefined macros as zero. We use compile option -Werror=undef which forbids such uglyness.

**Comment by sverker** [comment]
I added som more fixes for LibreSSL and rebased on a fresh maint branch.

**Comment by sverker** [comment]
Merged to maint for OTP 26.1.

**Pull request merged**